bin/runtime/src/lib.rs

```diff
@@ -1589,4 +1590,155 @@ mod tests {
 
         assert!(lhs < rhs);
     }
+
+    #[test]
+    fn era_payout() {
+        use sp_io::TestExternalities;
+
+        struct Inputs {
+            azero_cap: Balance,
+            horizon: u64,
+            // will be ignored
+            total_staked: Balance,
+            total_issuance: Balance,
+            era_duration_millis: u64,
+        }
+
+        struct Outputs {
+            validators_payout: Balance,
+            rest: Balance,
+        }
+
+        fn assert_on_data(inputs: Inputs, outputs: Outputs) {
+            TestExternalities::default().execute_with(|| {
+                pallet_aleph::AzeroCap::<Runtime>::put(inputs.azero_cap);
+                pallet_aleph::ExponentialInflationHorizon::<Runtime>::put(inputs.horizon);
+                let (validators_payout, rest) =
+                    <Runtime as pallet_staking::Config>::EraPayout::era_payout(
+                        inputs.total_staked,
+                        inputs.total_issuance,
+                        inputs.era_duration_millis,
+                    );
+                assert_eq!(validators_payout, outputs.validators_payout);
+                assert_eq!(rest, outputs.rest);
+            });
+        }
+
+        fn run_for_n_eras(inputs: Inputs, n_eras: usize) -> (Vec<Outputs>, Balance) {
+            let mut outputs = vec![];
+            let mut total_issuance = inputs.total_issuance;
+            for _ in 0..n_eras {
+                TestExternalities::default().execute_with(|| {
+                    pallet_aleph::AzeroCap::<Runtime>::put(inputs.azero_cap);
+                    pallet_aleph::ExponentialInflationHorizon::<Runtime>::put(inputs.horizon);
+                    let (validators_payout, rest) =
+                        <Runtime as pallet_staking::Config>::EraPayout::era_payout(
+                            inputs.total_staked,
+                            total_issuance,
+                            inputs.era_duration_millis,
+                        );
+                    outputs.push(Outputs {
+                        validators_payout,
+                        rest,
+                    });
+                    total_issuance += validators_payout + rest;
+                });
+            }
+            (outputs, total_issuance)
+        }
+
+        const MILLISECS_PER_DAY: u64 = 24 * 60 * 60 * 1000;
+
+        // standard case
+        assert_on_data(
+            Inputs {
+                azero_cap: 100_000_000 * TOKEN,
+                horizon: 365 * MILLISECS_PER_DAY,
+                total_staked: 0,
+                total_issuance: 50_000_000 * TOKEN,
+                era_duration_millis: MILLISECS_PER_DAY,
+            },
+            Outputs {
+                validators_payout: 123_118 * TOKEN + 920_000_000_000,
+                rest: 13_679 * TOKEN + 880_000_000_000,
+            },
+        );
+
+        // After 3 * horizon milliseconds the gap should be reduced by ~95%
```

Comment:
Because (1/e)^3 = 0.05? Is this written somewhere that for 1 horizon we expect 1/e reduction?

Reply:
Yup. And there's one check for a scenario in which era's length is equal to the horizon.